src/impurity.jl

```diff
@@ -32,135 +32,205 @@ function SolidStateDetectors.get_impurity_density(
+with David Radford's empirical function a + b*z + c*exp((z-L)/tau)
+where z is crystal coordinate from seed end (z=0) to crystal length(z=L)
 
+The fit is performed in crystal metadata coordinates
+i.e. distance in mm and impurity values in 1e9 e/cm^3
+which determines the units of the parameters a,b,c,tau returned in the vector of Float32.
+"""
+function impurity_density_model(crystal_metadata::PropDict)
     # get impurity measurements
     T = Float32
     # distance from seed end in mm (crystal axis)
-    dist = T.(crystal_dict.impurity_measurements.distance_from_seed_end_mm) # in mm for now
+    dist = T.(crystal_metadata.impurity_measurements.distance_from_seed_end_mm)
     # measurements corr to dist
-    imp_values = T.(crystal_dict.impurity_measurements.value_in_1e9e_cm3).* (1e6 * 1e9) # SSD in e/m^3
-
-    # position of detector axis Z=0 (p+ contact) of this detector (slice) in the crystal from seed end
-    det_z0 = crystal_dict.slices[Symbol(meta.production.slice)].detector_offset_in_mm
+    imp_values = T.(crystal_metadata.impurity_measurements.value_in_1e9e_cm3) #  .* (1e6 * 1e9) # SSD in e/m^3
 
-    # quadratic fit
+    # old: quadratic fit
     # convert points from crystal axis to detector axis, convert to SSD unit
     # dist_det = (det_z0 .- dist) ./ 1e3 # SSD in m
     # a, b, c = poly_fit(dist_det, imp_values, 2)
     # QuadraticImpurityDensity{Float32}(a, b, c)
 
-    # Radford fit -> in the future need to make settings to choose
-    # fit in crystal coordinates
-    # convert to SSD units
-    dist = dist / 1e3 #m
-    L = dist[end] - dist[1]
+    # Radford fit -> in the future need to make settings to choose?
+    # fit in crystal metadata coordinates
+    L = dist[end] - dist[1] # in mm
 
     @. radford(z, p) = p[1] + p[2] * z + p[3] * exp.((z-L)/p[4])
 
-    # @ z = 0 equals first value
+    ## starting values for fit
+    # f(z = 0) = a
     a = imp_values[1]
-    # 0.04 was a good starting value using mm and 1e9e/cm^3 -> scale
-    b = 0.04 * (1e6 * 1e9) * 1e3
-    # @ z = L
+    # 0.04 good starting value in mm and 1e9e/cm^3
+    b = 0.04 # * 1e3 * (1e6 * 1e9)
+    # f(z = L) = a + bL + c
     c = (imp_values[end] - a - b*L)
-    # tau = 20 was a good starting value in mm -> scale
-    tau = 20 / 1e3
+    # tau = 20 good starting value in mm
+    tau = 20 # / 1e3
     p0 = [a, b, c, tau]
+
+    ### fit in 1e9 e/cm^3 VS mm
     fit = LsqFit.curve_fit(radford, dist, imp_values, p0)
 
-    a,b,c,tau = fit.param
-    RadfordImpurityDensity(T(a),T(b),T(c),T(tau),T(L),T(det_z0/1e3))
+    fit.param
+end
+
+
+function impurity_density_model(meta::PropDict, crystal_metadata_path::AbstractString, simulator::PSSimulator)
+
+    # get crystal metadata
+    crystal_name = first(meta.name, length(meta.name)-1)
+    crystal_path = joinpath(crystal_metadata_path, crystal_name*".json")
+
+    if !ispath(crystal_path)
+        @error "Crystal $(meta.production.crystal) does not exist in path $crystal_path"
+    end
```

Comment:
To me, it seems like `crystal_metadata_path` is the path and `crystal_path` is the name of the file?
If that is correct, this should be 
```julia
if !isdir(crystal_metadata_path) 
    @error "Crystal metadata path $crystal_metadata_path does not exist"
end
if !isfile(crystal_path)
    @error "Crystal $(meta.production.crystal) does not exist in path $crystal_metadata_path"
end
```

Reply:
The check if crystal metadata folder path exists already happens in `pss.jl` in where the path is read from input settings. At that point the detector name and therefore crystal is not known to the simulator, so I only check the existence of the directory. Then in `imputiry.jl` I check the file itself. This is definitely a bad implementation, also because there's a copy-paste for the check in `SSDSimulator` and `SiggenSimulator`. I will do it the way you wrote - check path and then file in `impurity.jl`, and remove those check from the Simulators.